packages/contracts-bedrock/src/interop/L2CrossDomainMessenger.sol

```diff
@@ -0,0 +1,147 @@
+            return;
+        }
+
+        require(_destination != CHAIN_ID, "message cant be sent to self");
+
+        bytes memory data = abi.encodeWithSelector(
+            this.relayMessage.selector, messageNonce(), CHAIN_ID, msg.sender, _target, msg.value, _minGasLimit, _message
+        );
+
+        // (1) Send to the CrossL2Inbox.
+        // ***Note*** With an updated CrossDomainMessenger message format, this contract can replace the L2CDM predeploy
+        // and switch between L2ToL1MessagePasser/CrossL2Inbox based on `_destination`.
+        //
+        // i.e: CrossL2Inbox.sendMessage(source, destination, _minGasLimit, msg.value, data) // ensure _minGaslimit covers the base
+
+        // (2) Emit Events
+
+        unchecked {
+            ++msgNonce;
+        }
+    }
+
+    function relayMessage(
+        uint256 _nonce,
+        bytes32 _source,
+        address _sender,
+        address _target,
+        uint256 _value,
+        uint256 _minGasLimit,
+        bytes calldata _message
+    ) external payable {
+        (, uint16 version) = Encoding.decodeVersionedNonce(_nonce);
+        require(version == INTEROP_MESSAGE_VERSION, "message is not of the right version");
+
+        bytes32 messageHash = hashCrossDomainMessageV2(_nonce, _source, CHAIN_ID, _sender, _target, _value, _minGasLimit, _message);
+        require(successfulMessages[messageHash] == false, "message already processed");
+
+        // (1) Allow for replay
+        if (_sender == address(this)) {
+            assert(msg.value == _value);
+            assert(failedMessages[messageHash] == false);
+        } else {
+            require(msg.value == 0, "cannot replay with more funds");
+            require(failedMessages[messageHash], "message cannot be replayed");
+        }
+
+        // .. check for min gas, unsafe target, etc.
+
+        // (2) Relay Message
+        // -- Ignoring the reserved gas subtracted from gasleft() & xDomainMsgSender update
+        xDomainMsgSender = _sender;
+        bool success = SafeCall.call(_target, gasleft(), _value, _message);
+        xDomainMsgSender = Constants.DEFAULT_L2_SENDER;
+        if (success) {
+            successfulMessages[messageHash] = true;
+            // (3) Emit event
+        } else {
+            failedMessages[messageHash] = true;
+            // (3) Emit event
+        }
+    }
+
+    
+    /**
+     * Encoding.sol: update for this new message version
+     */
+
+    function encodeCrossDomainMessageV2(
+        uint256 _nonce,
+        bytes32 _source,
+        bytes32 _destination,
+        address _sender,
+        address _target,
+        uint256 _value,
+        uint256 _gasLimit,
+        bytes memory _data
+    ) internal pure returns (bytes memory) {
+        return abi.encodeWithSignature(
+            "relayMessage(uint256,bytes32,bytes32,address,address,uint256,uint256,bytes)",
```

Comment:
The encode and hash functions should live in the libraries `Encoding` and `Hashing` so they are portable

Reply:
Since we dont want to touch the core contracts outside of `/interop` i'm just including them in this file with a block comment that these belong in `Encoding` and `Hashing` for clarity